From f0039a73272fc3401ee8492aac01d02c24650ba4 Mon Sep 17 00:00:00 2001
From: Aihua Xu <aihuaxu@apache.org>
Date: Thu, 30 Mar 2017 14:08:08 -0400
Subject: [PATCH 1057/1164] CDH-51146: HIVE-16057: SchemaTool ignores
 --passWord argument if
 hadoop.security.credential.provider.path is
 configured (Peter Vary, reviewed by Aihua Xu)

(cherry picked from commit e339d144425f01bad2aaab1367bded15f5dbaaba)

Change-Id: I93acf499d43b1dafccd67d396bcde4aa88878ba3
---
 .../org/apache/hive/beeline/HiveSchemaTool.java    |   17 ++++++++++-------
 .../org/apache/hive/beeline/TestSchemaTool.java    |    5 +++++
 2 files changed, 15 insertions(+), 7 deletions(-)

diff --git a/beeline/src/java/org/apache/hive/beeline/HiveSchemaTool.java b/beeline/src/java/org/apache/hive/beeline/HiveSchemaTool.java
index 59f5786..e71da03 100644
--- a/beeline/src/java/org/apache/hive/beeline/HiveSchemaTool.java
+++ b/beeline/src/java/org/apache/hive/beeline/HiveSchemaTool.java
@@ -89,13 +89,6 @@ public HiveSchemaTool(String hiveHome, HiveConf hiveConf, String dbType)
     this.hiveConf = hiveConf;
     this.dbType = dbType;
     this.metaStoreSchemaInfo = new MetaStoreSchemaInfo(hiveHome, hiveConf, dbType);
-    userName = hiveConf.get(ConfVars.METASTORE_CONNECTION_USER_NAME.varname);
-    try {
-      passWord = ShimLoader.getHadoopShims().getPassword(hiveConf,
-          HiveConf.ConfVars.METASTOREPWD.varname);
-    } catch (IOException err) {
-      throw new HiveMetaException("Error getting metastore password", err);
-    }
   }
 
   public HiveConf getHiveConf() {
@@ -1118,9 +1111,19 @@ public static void main(String[] args) {
 
       if (line.hasOption("userName")) {
         schemaTool.setUserName(line.getOptionValue("userName"));
+      } else {
+        schemaTool.setUserName(
+            schemaTool.getHiveConf().get(ConfVars.METASTORE_CONNECTION_USER_NAME.varname));
       }
       if (line.hasOption("passWord")) {
         schemaTool.setPassWord(line.getOptionValue("passWord"));
+      } else {
+        try {
+          schemaTool.setPassWord(ShimLoader.getHadoopShims().getPassword(schemaTool.getHiveConf(),
+              HiveConf.ConfVars.METASTOREPWD.varname));
+        } catch (IOException err) {
+          throw new HiveMetaException("Error getting metastore password", err);
+        }
       }
       if (line.hasOption("dryRun")) {
         schemaTool.setDryRun(true);
diff --git a/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestSchemaTool.java b/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestSchemaTool.java
index 534d1de..95ba635 100644
--- a/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestSchemaTool.java
+++ b/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestSchemaTool.java
@@ -36,6 +36,7 @@
 import org.apache.hadoop.hive.conf.HiveConf;
 import org.apache.hadoop.hive.metastore.HiveMetaException;
 import org.apache.hadoop.hive.metastore.MetaStoreSchemaInfo;
+import org.apache.hadoop.hive.shims.ShimLoader;
 import org.apache.hive.beeline.HiveSchemaHelper.NestedScriptParser;
 import org.apache.hive.beeline.HiveSchemaHelper.PostgresCommandParser;
 
@@ -57,6 +58,10 @@ protected void setUp() throws Exception {
     hiveConf = new HiveConf(this.getClass());
     schemaTool = new HiveSchemaTool(
         System.getProperty("test.tmp.dir", "target/tmp"), hiveConf, "derby");
+    schemaTool.setUserName(
+        schemaTool.getHiveConf().get(HiveConf.ConfVars.METASTORE_CONNECTION_USER_NAME.varname));
+    schemaTool.setPassWord(ShimLoader.getHadoopShims().getPassword(schemaTool.getHiveConf(),
+          HiveConf.ConfVars.METASTOREPWD.varname));
     System.setProperty("beeLine.system.exit", "true");
     errStream = System.err;
     outStream = System.out;
-- 
1.7.9.5

